File: algorithms/68.cpp
#include <iostream>
#include <sstream>
#include <vector>
using namespace std;

class Solution {
private:
    string helper(const vector<string> &line, int length, int maxWidth, bool isLast) {
        int nspace, ext;
        if (line.size() == 1 || isLast) {
            nspace = 1;
            ext = 0;
        }
        else {
            nspace = (maxWidth-length) / (line.size()-1);
            ext = (maxWidth-length) % (line.size()-1);
        }
        stringstream ss;
        for (int i = 0; i < line.size(); i++) {
            ss << line[i];
            if (i != line.size()-1) {
                for (int j = 0; j < nspace; j++)
                    ss << " ";
                if (i < ext)
                    ss << " ";
            }
        }
        string result = ss.str();
        while (result.size() < maxWidth)
            result.push_back(' ');
        return result;
    }
public:
    vector<string> fullJustify(vector<string> &words, int maxWidth) {
        vector<string> result;
        vector<string> line;
        int length = 0;
        for (auto word: words) {
            if (word.size() + length + line.size() > maxWidth) {
                result.push_back(helper(line, length, maxWidth, false));
                line.clear();
                length = 0;
            }
            line.push_back(word);
            length += word.size();
        }
        if (line.size() > 0)
            result.push_back(helper(line, length, maxWidth, true));
        return result;
    }
};

int main() {
    return 0;
}

File: algorithms/rust/68.rs
impl Solution {
    #[allow(dead_code)]
    pub fn full_justify(words: Vec<String>, max_width: i32) -> Vec<String> {
        let max_width = max_width as usize;
        let mut rslt = vec![];

        let mut words_of_line = vec![];
        let mut width = 0;

        for word in words.iter() {
            if word.len() + words_of_line.len() + width > max_width {
                let mut line = String::new();

                if words_of_line.len() == 1 {
                    line.push_str(words_of_line[0]);
                    line.push_str(&" ".repeat(max_width - width));
                } else {
                    let sep = words_of_line.len() - 1;
                    let space_each_sep = (max_width - width) / sep;
                    let sep_has_extra_space = (max_width - width) % sep;
                    for (idx, &w) in words_of_line.iter().enumerate() {
                        if idx != 0 {
                            line.push_str(&" ".repeat(space_each_sep));
                            if idx <= sep_has_extra_space {
                                line.push(' ');
                            }
                        }
                        line.push_str(w);
                    }
                }
                words_of_line.clear();
                width = 0;
                rslt.push(line);
            }
            width += word.len();
            words_of_line.push(word);
        }

        let mut line = words_of_line.join(" ");
        line.push_str(&" ".repeat(max_width - line.len()));
        rslt.push(line);

        rslt
    }
}

struct Solution;

fn main() {
    println!("Hello, world!");
}
